python/thunder/utils/serializable.py

```diff
@@ -0,0 +1,410 @@
+                        valSer = self.__serializeRecursively(valAttrDict, numpyStorage)
+                        outData.append((keySer, valSer))
+                else:
+                    outData = [(self.__serializeRecursively(k, numpyStorage),
+                                self.__serializeRecursively(val.__dict__, numpyStorage))
+                               for (k, val) in data.iteritems()]
+                return {"py/hmgDict": {
+                    "type": valueType.__name__,
+                    "module": valueType.__module__,
+                    "data": outData
+                    }
+                }
+            elif all(type(k) == str for k in data):  # string keys can be represented natively in JSON
+                # avoid dict comprehension for py2.6 compatibility
+                retVal = {}
+                for k, v in data.iteritems():
+                    retVal[k] = self.__serializeRecursively(v, numpyStorage)
+                return retVal
+            else:
+                return {"py/dict": [(self.__serializeRecursively(k, numpyStorage),
+                                     self.__serializeRecursively(v, numpyStorage)) for k, v in data.iteritems()]}
+        elif dataType == tuple:                          # Recurse into tuples
+            return {"py/tuple": [self.__serializeRecursively(val, numpyStorage) for val in data]}
+        elif dataType == set:                            # Recurse into sets
+            return {"py/set": [self.__serializeRecursively(val, numpyStorage) for val in data]}
+        elif dataType == datetime.datetime:
+            return {"py/datetime.datetime": str(data)}
+        elif dataType == complex:
+            return {"py/complex": [data.real, data.imag]}
+        elif dataType == ndarray:
+            if numpyStorage == 'ascii' or (numpyStorage == 'auto' and data.size < 1000):
+                return {"py/numpy.ndarray": {
+                    "encoding": "ascii",
+                    "shape": data.shape,
+                    "values": data.tolist(),
+                    "dtype":  str(data.dtype)}}
+            else:
+                from base64 import b64encode
+                return {"py/numpy.ndarray": {
+                    "encoding": "base64",
+                    "shape": data.shape,
+                    "values": b64encode(data),
+                    "dtype":  str(data.dtype)}}
+        elif isinstance(data, Serializable):
+            # nested serializable object
+            return {"py/Serializable": {
+                "type": dataType.__name__,
+                "module": dataType.__module__,
+                "data": data.serialize()
+            }}
+
+        raise TypeError("Type %s not data-serializable" % dataType)
+
+    def serialize(self, numpyStorage='auto'):
+        """
+        Serialize this object to a python dictionary that can easily be converted
+        to/from JSON using Python's standard JSON library.
```

Comment:
We're using all caps for `JSON` in the docstrings, so might make sense to do it that way for the to-be-added `toJSON` and `fromJSON` methods.